cocotb/binary.py

```diff
@@ -325,36 +330,41 @@ def is_resolvable(self):
         return not any(char in self._str for char in BinaryValue._resolve_to_error)
 
     @property
-    def buff(self):
-        """Attribute :attr:`buff` represents the value as a binary string buffer.
+    def buff(self) -> bytes:
+        r"""The value as a binary string buffer.
 
-        >>> "0100000100101111".buff == "\x41\x2F"
+        >>> BinaryValue("01000001" + "00101111").buff == b"\x41\x2F"
         True
+
+        .. versionchanged:: 1.4
+            This changed from :class:`str` to :class:`bytes`.
+            Note that for older versions used with Python 2 these types were
+            indistinguishable.
         """
         bits = resolve(self._str)
 
         if len(bits) % 8:
             bits = "0" * (8 - len(bits) % 8) + bits
 
-        buff = ""
+        buff = []
         while bits:
             byte = bits[:8]
             bits = bits[8:]
             val = int(byte, 2)
             if self.big_endian:
-                buff += chr(val)
+                buff += [val]
             else:
-                buff = chr(val) + buff
-        return buff
+                buff = [val] + buff
+        return bytes(buff)
```

Comment:
Alternatively,

```python
n_bytes = ceil(len(bits) / 8)
val = int(bits, 2)
return val.to_bytes(n_bytes, byteorder=('big' if self.big_endian else 'little'))
```

Reply:
I'd probably want to leave that to a follow-up. I'm not 100% convinced that's the same, and would need to write some tests to be sure.

I also wonder if there's a way to avoid using the intermediate arbitrary-precision integer.

Reply:
Also, a neat trick you might enjoy that doesn't need floating point numbers - `n_bytes = -(-bits // 8)`